File: backend/app/services/service.py
from uuid import UUID
from typing import Optional, List
from app.models import ServiceInDB, ServiceCreate, ServiceUpdate, ServiceResponse
from app.repository import ServiceRepository
from app.utils.exceptions import (
    NotFoundException,
    InternalServerException,
    ConfictException,
)


def service_in_db_to_response(service: ServiceInDB) -> ServiceResponse:
    """"""
    return ServiceResponse(**service.model_dump())


class ServiceService:
    def __init__(self, repo: ServiceRepository) -> None:
        self.repo = repo

    async def create_service(self, data: ServiceCreate) -> ServiceResponse:
        """"""
        if await self.repo.exists(data.name):
            raise ConfictException(f"{data.name} service already exists")
        service: Optional[ServiceInDB] = await self.repo.create(data.model_dump())
        if service is None:
            raise InternalServerException("Error creating service")
        return service_in_db_to_response(service)

    async def readone_service(self, service_id: UUID) -> ServiceResponse:
        """"""
        service: Optional[ServiceInDB] = await self.repo.readone(service_id)
        if service is None:
            raise NotFoundException("Service not found")
        return service_in_db_to_response(service)

    async def readall_services(
        self, skip: int = 0, limit: int = 100
    ) -> List[ServiceResponse]:
        """"""
        services: List[ServiceInDB] = await self.repo.readall(skip, limit)
        return [service_in_db_to_response(service) for service in services]

    async def update_service(
        self, service_id: UUID, data: ServiceUpdate
    ) -> ServiceResponse:
        """"""
        service: Optional[ServiceInDB] = await self.repo.readone(service_id)
        if service is None:
            raise NotFoundException("Service not found")
        service: Optional[ServiceInDB] = await self.repo.update(
            service_id, data.model_dump()
        )
        if service is None:
            raise InternalServerException("Error updating service")
        return service_in_db_to_response(service)

    async def delete_service(self, service_id: UUID) -> bool:
        """"""
        return await self.repo.delete(service_id)

    async def readone_service_by_name(self, service_name: str) -> ServiceResponse:
        """"""
        service: Optional[ServiceInDB] = await self.repo.readone_by_name(service_name)
        if service is None:
            raise NotFoundException(f"Service {service_name} not found")
        return service_in_db_to_response(service)

File: frontend/src/lib/services/service.ts
import axios from "axios";
import { BASE_API_URL } from "@/lib/constants";
import type { ServiceCreate, ServiceResponse, ServiceUpdate } from "../types/service";

const URL = `${BASE_API_URL}/service`;

export async function createService(data: ServiceCreate): Promise<ServiceResponse> {
    const response = await axios.post(URL, data);
    return response.data;
}

export async function readoneService(serviceID: string): Promise<ServiceResponse> {
    const response = await axios.get(`${URL}/${serviceID}`);
    return response.data;
}

export async function readallServices(
    skip: number = 0,
    limit: number = 100
): Promise<ServiceResponse[]> {
    const params = {skip: skip, limit: limit};
    const response = await axios.get(URL, {params});
    return response.data;
}

export async function updateService(
    serviceID: string,
    data: ServiceUpdate
): Promise<ServiceResponse> {
    const response = await axios.put(`${URL}/${serviceID}`, data);
    return response.data;
}

export async function deleteService(serviceID: string): Promise<boolean> {
    const response = await axios.delete(`${URL}/${serviceID}`);
    return response.data;
}

export async function readoneServiceByName(serviceName: string): Promise<ServiceResponse> {
    const response = await axios.get(`${URL}/lookup/${serviceName}`);
    return response.data;
}
